src/builder.ts

```diff
@@ -0,0 +1,70 @@
+import { resolveURL, withQuery } from 'ufo'
+import { defu } from 'defu'
+import type { QueryObject } from 'ufo'
+import type { FetchOptions } from './fetch'
+import type { ResponseType, MappedType } from './types'
+import { $fetch } from './undici'
+
+export type ClientMethodHandler = <T = any, R extends ResponseType = 'json'>(
+  data?: RequestInit['body'] | Record<string, any>,
+  opts?: Omit<FetchOptions<R>, 'baseURL' | 'method'>
+) => Promise<MappedType<R, T>>
+
+export type ClientBuilder = {
+  [key: string]: ClientBuilder;
+  (...segmentsOrIds: (string | number)[]): ClientBuilder
+} & {
+  get: ClientMethodHandler
+  post: ClientMethodHandler
+  put: ClientMethodHandler
+  delete: ClientMethodHandler
+  patch: ClientMethodHandler
+}
+
+export function createClient<R extends ResponseType = 'json'> (
+  url: string,
+  defaults: Omit<FetchOptions<R>, 'method'> = {}
+): ClientBuilder {
+  // Callable internal target required to use `apply` on it
+  const internalTarget = (() => {}) as ClientBuilder
+
+  function p (url: string): ClientBuilder {
+    return new Proxy(internalTarget, {
+      get (_target, key: string) {
+        const method = key.toUpperCase()
+
+        if (!['GET', 'POST', 'PUT', 'DELETE', 'PATCH'].includes(method)) {
+          return p(resolveURL(url, key))
+        }
+
+        const handler: ClientMethodHandler = <T = any, R extends ResponseType = 'json'>(
+          data?: RequestInit['body'] | Record<string, any>,
+          opts: FetchOptions<R> = {}
+        ) => {
+          switch (method) {
+            case 'GET':
+              if (data) {
+                url = withQuery(url, data as QueryObject)
+              }
+              break
+            case 'POST':
+            case 'PUT':
+            case 'PATCH':
+              opts.body = data
+          }
+
+          opts.method = method
+
+          return $fetch<T, R>(url, defu(opts, defaults) as FetchOptions<R>)
```

Comment:
As long as options are flat, we probably don't need `defu` and can use `{ ...defaults, ...opts }`

Reply:
That was my initial idea as well. But what about extending `headers`? We could extend them on a request method level with defu.

Reply:
Good point about headers! One issue with header merging is that the user can potentially pass a [Headers](https://developer.mozilla.org/en-US/docs/Web/API/Headers) class as well and well that
 won't be merged as expected (until we [normalize `headers`](https://github.com/unjs/ohmyfetch/blob/26641a4405826de6281e357bdd665df692d0f340/src/fetch.ts#L110)). What about introducing new `defaultHeaders` global option?

Reply:
Right, but wouldn't a new option like `defaultHeaders` have to merged as well? I think normalizing headers is the more semantic approach, rather than introducing a new option. Let's postpone that and don't extend the scope of this PR too much. I think another PR focusing on normalizing headers would be cleaner. 🙂 

I have dropped defu and will introduce header merging later. Would do you say?

Reply:
Merging headers is supported since https://github.com/unjs/ofetch/pull/69/commits/41e267ec4b1b02d072053078e5bd4db203f0a091.